File: 프로그래머스/3/12927. 야근 지수/야근 지수.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

long long solution(int n, vector<int> works) {
    priority_queue<int> pq;
    for(int w : works) pq.push(w);
    
    while(n-->0 && !pq.empty()){
        int tmp = pq.top();
        pq.pop();
        if(tmp>0) pq.push(--tmp);
    }
    
    long long res = 0;
    while(!pq.empty()){
        long long num = pq.top();
        res += num * num;
        pq.pop();
    }
    return res;
}

File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.*;

class Solution {
    public long solution(int n, int[] works) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        for (int work : works) {
            pq.offer(work);
        }
        while(n-->0 && !pq.isEmpty()){
            int tmp = pq.poll();
            if(tmp>0) pq.offer(--tmp);
        }
        
        long res=0;
        while(!pq.isEmpty()){
            res += (long) Math.pow(pq.poll(), 2);
        }
        return res;
    }
}